test/functional/p2p_add_connections.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test add_outbound_p2p_connection test framework functionality"""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.p2p import P2PInterface
+from test_framework.util import assert_equal
+
+
+class P2PAddConnections(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = False
+        self.num_nodes = 2
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def check_node_connections(self, node=None, num_in=None, num_out=None):
+        info = self.nodes[node].getnetworkinfo()
+        assert_equal(info["connections_in"], num_in)
+        assert_equal(info["connections_out"], num_out)
+
+    def run_test(self):
+        self.disconnect_nodes(0, 1)
+
+        self.log.info("Add 8 outbounds to node 0")
+        for i in range(8):
+            self.log.info("outbound: {}".format(i))
+            self.nodes[0].add_outbound_p2p_connection(P2PInterface(), connection_type="outbound")
+
+        self.log.info("Add 2 block-relay-only connections to node 0")
+        for i in range(2):
+            self.log.info("block-relay-only: {}".format(i))
+            self.nodes[0].add_outbound_p2p_connection(P2PInterface(), connection_type="blockrelay")
+
+        self.log.info("Add 2 block-relay-only connections to node 1")
+        for i in range(2):
+            self.log.info("block-relay-only: {}".format(i))
+            self.nodes[1].add_outbound_p2p_connection(P2PInterface(), connection_type="blockrelay")
+
+        self.log.info("Add some inbound connections to node 1")
+        for i in range(5):
+            self.log.info("inbound: {}".format(i))
+            self.nodes[1].add_p2p_connection(P2PInterface())
+
+        self.log.info("Add 8 outbounds to node 1")
+        for i in range(8):
+            self.log.info("outbound: {}".format(i))
+            self.nodes[1].add_outbound_p2p_connection(P2PInterface())
+
+        self.log.info("Check the connections opened as expected")
+        self.check_node_connections(node=0, num_in=0, num_out=10)
+        self.check_node_connections(node=1, num_in=5, num_out=10)
+
+        self.log.info("Disconnect p2p connections & try to re-open")
+        self.nodes[0].disconnect_p2ps()
+        self.check_node_connections(node=0, num_in=0, num_out=0)
+
+        self.log.info("Add 8 outbounds to node 0")
+        for i in range(8):
+            self.log.info("outbound: {}".format(i))
```

Comment:
Oh, now that the minimum Python version has been bumped to 3.6 (in #19504), you can use Python f-strings instead of `format` for new code.

Reply:
done